vtm/src/org/oscim/tiling/source/mapfile/MapDatabase.java

```diff
@@ -434,20 +430,30 @@ private void setTileClipping(QueryParameters queryParameters, long mCurrentRow,
         int ySmax = Tile.SIZE;
 
         if (numRows > 0) {
-            int w = (int) (Tile.SIZE / (numCols + 1));
-            int h = (int) (Tile.SIZE / (numRows + 1));
-
-            if (mCurrentCol > 0)
-                xSmin = xmin = (int) (mCurrentCol * w);
-
-            if (mCurrentCol < numCols)
-                xSmax = xmax = (int) (mCurrentCol * w + w);
-
-            if (mCurrentRow > 0)
-                ySmin = ymin = (int) (mCurrentRow * h);
-
-            if (mCurrentRow < numRows)
-                ySmax = ymax = (int) (mCurrentRow * h + h);
+            /* If blocks are at a border, sometimes too less blocks are requested,
+             * so the divisor for tile dimensions is increased to base tile subdivision.
+             */
+            boolean isTopBorder = subFileParameter.boundaryTileTop > queryParameters.fromBaseTileY;
+            boolean isLeftBorder = subFileParameter.boundaryTileLeft > queryParameters.fromBaseTileX;
+            int numSubX = (int) (queryParameters.toBaseTileX - queryParameters.fromBaseTileX) + 1;
+            int numSubY = (int) (queryParameters.toBaseTileY - queryParameters.fromBaseTileY) + 1;
+            int numDifX = numSubX - (numCols + 1); // Is 0, except at map borders
+            int numDifY = numSubY - (numRows + 1); // Is 0, except at map borders
+
+            int w = Tile.SIZE / numSubX;
+            int h = Tile.SIZE / numSubY;
+
+            if (currentCol > 0)
+                xSmin = xmin = (int) ((currentCol + (isLeftBorder ? numDifX : 0)) * w);
+
+            if (currentCol < numCols)
+                xSmax = xmax = (int) ((currentCol + (isLeftBorder ? numDifX : 0)) * w + w);
+
+            if (currentRow > 0)
+                ySmin = ymin = (int) ((currentRow + (isTopBorder ? numDifY : 0)) * h);
+
+            if (currentRow < numRows)
+                ySmax = ymax = (int) ((currentRow + (isTopBorder ? numDifY : 0)) * h + h);
```

Comment:
The 4 border checks seem not needed, since numDifX/Y already are or not 0?

Reply:
Even if `numDifX` isn't 0, at the right/bottom border it shouldn't be added as there the last block is missing instead of the first block.